packages/firestore/src/model/document.ts

```diff
@@ -100,85 +207,68 @@ export class Document extends MaybeDocument {
   }
 
   toString(): string {
-    return (
-      `Document(${this.key}, ${this.version}, ${this.data.toString()}, ` +
-      `{hasLocalMutations: ${this.hasLocalMutations}}), ` +
-      `{hasCommittedMutations: ${this.hasCommittedMutations}})`
-    );
-  }
-
-  get hasPendingWrites(): boolean {
-    return this.hasLocalMutations || this.hasCommittedMutations;
-  }
-
-  static compareByField(field: FieldPath, d1: Document, d2: Document): number {
-    const v1 = d1.field(field);
-    const v2 = d2.field(field);
-    if (v1 !== undefined && v2 !== undefined) {
-      return v1.compareTo(v2);
-    } else {
-      return fail("Trying to compare documents on fields that don't exist");
+    switch (this.type) {
+      case DocumentType.UNKNOWN:
+        return `Document(UNKNOWN, ${this.key})`;
+      case DocumentType.CONTENTS_UNKNOWN:
+        return `Document(CONTENTS_UNKNOWN, ${this.key}, ${this.version})`;
+      case DocumentType.MISSING:
+        return `Document(MISSING, ${this.key}, ${this.version})`;
+      case DocumentType.EXISTS:
+        return (
+          `Document(${this.key}, ${this.version}, ${this.data.toString()}, ` +
+          `documentState=${this.documentState})`
+        );
+      default:
+        return fail('Unknown document type');
     }
   }
-}
-
-/**
- * A class representing a deleted document.
- * Version is set to 0 if we don't point to any specific time, otherwise it
- * denotes time we know it didn't exist at.
- */
-export class NoDocument extends MaybeDocument {
-  readonly hasCommittedMutations: boolean;
 
-  constructor(
-    key: DocumentKey,
-    version: SnapshotVersion,
-    options?: DocumentOptions
-  ) {
-    super(key, version);
-    this.hasCommittedMutations = !!(options && options.hasCommittedMutations);
+  get exists(): boolean {
+    return this.type === DocumentType.EXISTS;
```

Comment:
Naively I would think that CONTENTS_UNKNOWN should return true.  If it shouldn't, then perhaps `.exists` is misnamed and really means something else?